File: packages/sdk/src/opaque-client.ts
// Thin seam between the SDK's HTTP register/login choreography (ADR-0007
// §3: "the SDK owns the actual register/login HTTP choreography [...];
// packages/crypto-wasm only ever returns/consumes protocol message bytes
// and opaque state bytes, never makes an HTTP call itself") and the
// concrete WASM binding. Defined as an interface, not a direct
// `packages/crypto-wasm` import inside AuthClient, so tests can supply a
// lightweight fake and only the one integration test needs a real WASM
// build.

export interface OpaqueStartResult {
  message: Uint8Array;
  state: Uint8Array;
}

export interface OpaqueFinishResult {
  message: Uint8Array;
}

export interface OpaqueClient {
  clientRegistrationStart(password: Uint8Array): OpaqueStartResult;
  clientRegistrationFinish(
    state: Uint8Array,
    password: Uint8Array,
    response: Uint8Array,
  ): OpaqueFinishResult;
  clientLoginStart(password: Uint8Array): OpaqueStartResult;
  clientLoginFinish(
    state: Uint8Array,
    password: Uint8Array,
    response: Uint8Array,
    context: Uint8Array,
  ): OpaqueFinishResult;
}

/** The fixed OPAQUE application-context bytes the real backend uses
 * (`OPAQUE_CONTEXT` in `apps/backend/src/auth/routes.mjs`). Login fails
 * generically against the real server if any other bytes are used. */
export const OPAQUE_CONTEXT = new TextEncoder().encode("zkpm-opaque-v1");

/** Loads the production OpaqueClient backed by `packages/crypto-wasm`'s
 * generated browser WASM module (ADR-0007). Callers must `await` this
 * before use; it calls the module's `init()` internally. Not used by unit
 * tests that supply a fake OpaqueClient — only by the real end-to-end
 * integration test and by real client applications (e.g. C01). */
export async function createWasmOpaqueClient(): Promise<OpaqueClient> {
  const mod = await import("crypto-wasm");
  await mod.default();
  return {
    clientRegistrationStart: (password) => mod.client_registration_start(password) as OpaqueStartResult,
    clientRegistrationFinish: (state, password, response) =>
      mod.client_registration_finish(state, password, response) as OpaqueFinishResult,
    clientLoginStart: (password) => mod.client_login_start(password) as OpaqueStartResult,
    clientLoginFinish: (state, password, response, context) =>
      mod.client_login_finish(state, password, response, context) as OpaqueFinishResult,
  };
}

File: apps/android/app/src/main/kotlin/com/pass/android/crypto/OpaqueClient.kt
package com.pass.android.crypto

import uniffi.crypto_ffi.opaqueClientLoginFinish
import uniffi.crypto_ffi.opaqueClientLoginStart
import uniffi.crypto_ffi.opaqueClientRegistrationFinish
import uniffi.crypto_ffi.opaqueClientRegistrationStart

/** OPAQUE application context bytes: must match the server's exactly
 * (`OPAQUE_CONTEXT = "zkpm-opaque-v1"` in `apps/backend/src/auth/routes.mjs`)
 * or login fails generically. */
val OPAQUE_CONTEXT: ByteArray = "zkpm-opaque-v1".toByteArray(Charsets.UTF_8)

data class LoginStart(val message: ByteArray, val state: ByteArray)
data class LoginFinish(val message: ByteArray)
data class RegistrationStart(val message: ByteArray, val state: ByteArray)
data class RegistrationFinish(val message: ByteArray)

/** Extracted so `VaultManager`'s unit tests can inject a fake OPAQUE client
 * (no real handshake, so a `FakeVaultApi` can be scripted without also
 * standing up a real OPAQUE server) — `OpaqueClient` is the only real
 * implementation. */
interface OpaqueClientApi {
    fun clientLoginStart(password: ByteArray): LoginStart
    fun clientLoginFinish(state: ByteArray, password: ByteArray, response: ByteArray, context: ByteArray = OPAQUE_CONTEXT): LoginFinish
    fun clientRegistrationStart(password: ByteArray): RegistrationStart
    fun clientRegistrationFinish(state: ByteArray, password: ByteArray, response: ByteArray): RegistrationFinish
}

/**
 * Thin Kotlin-idiomatic wrapper over the generated `uniffi.crypto_ffi`
 * OPAQUE client bindings (D02-MVP's new crypto-ffi exports). Mirrors
 * `packages/sdk/src/opaque-client.ts`'s `OpaqueClient` interface shape;
 * `packages/sdk` is a forbidden path for this task so this is a deliberate
 * mirror. All protocol math happens inside Rust — this class never sees or
 * computes a key.
 */
class OpaqueClient : OpaqueClientApi {
    override fun clientLoginStart(password: ByteArray): LoginStart {
        val result = opaqueClientLoginStart(password)
        return LoginStart(result.message, result.state)
    }

    override fun clientLoginFinish(state: ByteArray, password: ByteArray, response: ByteArray, context: ByteArray): LoginFinish {
        val result = opaqueClientLoginFinish(state, password, response, context)
        return LoginFinish(result.message)
    }

    override fun clientRegistrationStart(password: ByteArray): RegistrationStart {
        val result = opaqueClientRegistrationStart(password)
        return RegistrationStart(result.message, result.state)
    }

    override fun clientRegistrationFinish(state: ByteArray, password: ByteArray, response: ByteArray): RegistrationFinish {
        val result = opaqueClientRegistrationFinish(state, password, response)
        return RegistrationFinish(result.message)
    }
}
